### PR TITLE
CIS-198 Reconnect Web Socket if no pong is received

### PR DESCRIPTION
**Pull request by buh**
- Added `WebSocketPingController`:
  - ping by repeating timer
  - checks if pong was received otherwise call force disconnect
  - resume/suspend ping timer when connection state did change.
- Added tests for `WebSocketPingController`.
- Integrated `WebSocketPingController` to `WebSocketClient` and removed internal ping timer.
- Updated `WebSocketClient` tests.

### CODE REVIEW

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+class WebSocketPingController {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:7]: Missing documentation. What does this class do?

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    typealias PingAction = () -> Void
+    typealias ForceReconnectAction = () -> Void
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:9]: I'm personally not a fan of type-aliasing simple callbacks. I find it more difficult to read because you don't from the function signature where it's used, what the signature is.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    
+    /// The time interval to ping connection to keep it alive.
+    static let pingTimeInterval: TimeInterval = 25
+    static let pongTimeoutTimeInterval: TimeInterval = 3
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:13]: Missing comment. What is this?

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    private let timerQueue: DispatchQueue
+    
+    /// The timer used for scheduling `ping` calls
+    private lazy var pingTimerControl: RepeatingTimerControl =
+        timerType.scheduleRepeating(timeInterval: Self.pingTimeInterval, queue: timerQueue) { [weak self] in self?.sendPing() }
+    
+    /// The pong timeout timer.
+    private var pongTimeoutTimer: TimerControl?
+    private let ping: () -> Void
+    private let forceReconnect: () -> Void
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:25]: There's no reason why this can't be `"internal" var`

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    /// Creates a ping controller.
+    /// - Parameters:
+    ///   - timerType: a timer type.
+    ///   - timerQueue: a timer dispatch queue.
+    ///   - ping: an action for `WebSocketClient` to send a ping.
+    ///   - forceReconnect: an action for `WebSocketClient` to force disconnect and reconnect.
+    init(timerType: Timer.Type,
+         timerQueue: DispatchQueue,
+         ping: @escaping PingAction,
+         forceReconnect: @escaping ForceReconnectAction) {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:36]: You will make fun of me, but why aren't these callbacks a simple delegate with two methods, that `WebSocketClient` implements? I think the API will be much nicer.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController_Tests.swift [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        time = VirtualTime()
+        VirtualTimeTimer.time = time
+        
+        pingCount = 0
+        reconnectCount = 0
+        
+        pingController = .init(timerType: VirtualTimeTimer.self,
+                               timerQueue: .main,
+                               ping: { [weak self] in self?.pingCount += 1 },
+                               forceReconnect: { [weak self] in self?.reconnectCount += 1 })
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController_Tests.swift:25]: You can/should use `unowned` here

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController_Tests.swift [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    }
+    
+    func test_ping() throws {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController_Tests.swift:28]: I think we need more tests. For example, you should test:
- that pings are sent periodically
- that pings are sent only when the connection is active
etc...

File: StreamChat.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -245,6 +246,8 @@
 		8A5B6A6124810C6B0082C023 /* Channel+KeystrokeTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8A5B6A6024810C6B0082C023 /* Channel+KeystrokeTests.swift */; };
 		8A5B6A6424810E660082C023 /* ClientTestCase.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8A5B6A6324810E660082C023 /* ClientTestCase.swift */; };
 		8A5D3EF924AF749200E2FE35 /* ChannelId_Tests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8A5D3EF824AF749200E2FE35 /* ChannelId_Tests.swift */; };
+		8A618E4524D19D510003D83C /* WebSocketPingController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8A618E4424D19D510003D83C /* WebSocketPingController.swift */; };
+		8A618E4824D2D1160003D83C /* WebSocketEventsMock.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8A9BC2EB2487D9F5004C5E69 /* WebSocketEventsMock.swift */; };
```
Comment (VojtaStavik) [inline StreamChat.xcodeproj/project.pbxproj:250]: Why is this here???

File: StreamChat.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -2901,6 +2910,7 @@
 				7988F6B124D010890004219B /* TestRunnerEnvironment.swift in Sources */,
 				799C9469247D791A001F1104 /* AssertResult.swift in Sources */,
 				8AC9CBD424C7351D006E236C /* ReactionEvents_Tests.swift in Sources */,
+				8A618E4824D2D1160003D83C /* WebSocketEventsMock.swift in Sources */,
```
Comment (VojtaStavik) [inline StreamChat.xcodeproj/project.pbxproj:2913]: and this???

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        self.timerQueue = timerQueue
+        self.ping = ping
+        self.forceReconnect = forceReconnect
+    }
+    
+    func connectionStateDidChange(_ connectionState: ConnectionState) {
+        cancelPongTimeoutTimer()
+        
+        if connectionState.isConnected {
+            log.info("🏓 Resume ping timer")
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:47]: Please don't use emojis in log. Our logger supports emojis in the formatter.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 11]
```diff
@@ -16,7 +13,7 @@ class WebSocketClient {
     }
     
     /// The notification center `WebSocketClient` uses to send notifications about incoming events.
-    private(set) lazy var notificationCenter: NotificationCenter = environment.notificationCenterBuilder()
+    private(set) lazy var notificationCenter: NotificationCenter = environment.createNotificationCenter()
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:16]: This should be in a separate commit.

File: StreamChatClient_v3/ChatClient.swift [hunk 12]
```diff
@@ -87,20 +87,21 @@ public class Client<ExtraData: ExtraDataTypes> {
...
+                                                                 encoder,
+                                                                 EventDecoder<ExtraData>(),
+                                                                 eventMiddlewares)
         
-        wsClient.connectionStateDelegate = self
+        webSocketClient.connectionStateDelegate = self
         
-        return wsClient
+        return webSocketClient
     }()
```
Comment (VojtaStavik) [inline StreamChatClient_v3/ChatClient.swift:105]: This should be a separate commit

File: StreamChatClient_v3/ChatClient.swift [hunk 13]
```diff
@@ -379,14 +380,17 @@ extension Client {
...
+                            requestEncoder: $2,
+                            eventDecoder: $3,
+                            eventMiddlewares: $4)
+        }
         
-        var databaseContainerBuilder: (_ kind: DatabaseContainer.Kind) throws
-            -> DatabaseContainer = { try DatabaseContainer(kind: $0) }
+        var databaseContainerBuilder: (_ kind: DatabaseContainer.Kind) throws -> DatabaseContainer = {
+            try DatabaseContainer(kind: $0)
+        }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/ChatClient.swift:394]: This also should be a separate commit since it has nothing to do with "Replace internal ping timer in WebSocketClient with WebSocketPingController"

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 14]
```diff
@@ -25,13 +22,11 @@ class WebSocketClient {
...
             if connectionState.isConnected {
-                pingTimer.resume()
                 reconnectionStrategy.sucessfullyConnected()
-                
-            } else {
-                pingTimer.suspend()
             }
             
+            pingController.connectionStateDidChange(connectionState)
+
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:29]: 👍

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 15]
```diff
@@ -94,6 +82,15 @@ class WebSocketClient {
...
     private let environment: Environment
     
+    private lazy var pingController: WebSocketPingController =
+        environment.createPingController(environment.timerType,
+                                         engineQueue,
+                                         { [weak self] in self?.engine.sendPing() },
+                                         { [weak self] in
+                                             self?.disconnect()
+                                             self?.reconnectIfNeeded(disconnectionError: ClientError.WebSocket("Pong timeout"))
+        })
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:92]: This should be `[unowned]`

Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:92]: Sorry, but I find this API not the most elegant one.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 16]
```diff
@@ -209,18 +206,32 @@ protocol ConnectionStateDelegate: AnyObject {
...
+        
+        var createNotificationCenter: () -> NotificationCenter = NotificationCenter.init
+        
+        var createEngine: CreateEngine = {
+            if #available(iOS 13, *) {
+                return URLSessionWebSocketEngine(request: $0, sessionConfiguration: $1, callbackQueue: $2)
+            } else {
+                return StarscreamWebSocketProvider(request: $0, sessionConfiguration: $1, callbackQueue: $2)
             }
+        }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:234]: You should rebase, this part has changes IIRC

File: StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift [hunk 17]
```diff
@@ -243,16 +244,42 @@ class WebSocketClient_Tests: StressTestCase {
...
+        time.run(numberOfSeconds: timeInterval)
+        XCTAssertEqual(engine.disconnect_calledCount, 1)
+        XCTAssertFalse(webSocketClient.connectionState.isConnected)
         
-        // Simulate time 3x longer than pingTimeInterval and assert 3 more pings
-        time.run(numberOfSeconds: 3 * pingInterval)
-        XCTAssertEqual(engine.sendPing_calledCount, 1 + 3)
+        AssertAsync {
+            Assert.willBeTrue(self.engine.connect_calledCount == 2)
+        }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift:282]: These tests should be here at all. You just need to test the integration, not the functionality:
- You should test that `pingController.connectionStateDidChange` is called every time the connection state changes.
- Test that when the controller calls ping the engine is called with ping
- Test that when pong event is received, `pingController.pongReceived()` is called

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 2 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:11]: I do that because these types are using outside.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 4 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift]: It prevents from inappropriate usage and any side effects. I don't think that the behaviour of the closure should be able to change later. Better to create another controller.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 5 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:41]: It's much more easy to use as I commented in our channel.

File: StreamChat.xcodeproj/project.pbxproj (hunk 8 above)
Comment (buh) [reply StreamChat.xcodeproj/project.pbxproj:250]: I added it to use `healthCheckEvent` data.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 10 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:52]: How I can setup emoji in our Logger for ping/pong only?

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 16 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:234]: What exactly changed?

File: Sources/Client/WebSocket/WebSocket.swift [hunk 18]
```diff
@@ -10,9 +10,6 @@ import UIKit
 
 /// A web socket client.
 final class WebSocket {
-    /// The time interval to ping connection to keep it alive.
-    static let pingTimeInterval: TimeInterval = 25
-
```
Comment (VojtaStavik) [inline Sources/Client/WebSocket/WebSocket.swift:15]: It seems you removed this from the v2? :)

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 2 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:11]: @b-onc , can you add your opinion? It's a tie :)

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 4 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift]: What exactly is the "inappropriate usage" and "side effect" of setting this callback? I'm not sure I understand.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 2 above)
Comment (b-onc) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:11]: I think it depends on the situation.
In the case of these 2, I agree that typealiasing makes using them harder, for the given reason (autocomplete doesn't show closure signature). And the typealias is not needed since the signature is not repeated.
I've also realized `ping` variable is of type `() -> Void` instead of `PingAction`, making the typealias totally useless in this case :) If the typealias is used extensively, then I'd support it, since typealiasing would save writing the signature everywhere.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 4 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift]: What I wanted to say is that it's only for internal usage. It shouldn't be used outside of PingController.
So this is not a desired flow:
```swift
pingController. forceReconnect()
```

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 2 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:11]: I use it like this:
```swift
        typealias CreatePingController = (
            _ timerType: Timer.Type,
            _ timerQueue: DispatchQueue,
            _ ping: @escaping WebSocketPingController.PingAction,
            _ forceReconnect: @escaping WebSocketPingController.ForceReconnectAction
        ) -> WebSocketPingController
```
```swift
    init(timerType: Timer.Type,
         timerQueue: DispatchQueue,
         ping: @escaping PingAction,
         forceReconnect: @escaping ForceReconnectAction) {
        self.timerType = timerType
        self.timerQueue = timerQueue
        self.ping = ping
        self.forceReconnect = forceReconnect
    }
```
I forgot to change it here:
```swift
    private let ping: PingAction
    private let forceReconnect: ForceReconnectAction
```

Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:11]: @b-onc do you think I should remove `typealias`?

Comment (b-onc) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:11]: For these 2, yes. Typealiasing closures are useful when closure signature is complex, and it's passed around many times

Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:11]: This seems to more like a workaround for the API design, as mentioned here https://github.com/GetStream/stream-chat-swift/pull/373#discussion_r463581266

Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift]: I agree, it's simple closure. I removed typealias

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift (hunk 10 above)
Comment (b-onc) [reply StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:49]: @buh 
```swift
public class PingPongEmojiFormatter: LogFormatter {
    public func format(logDetails: LogDetails, message: String) -> String {
        if logDetails.message.contains("WebSocket Ping") {
            return "🏓➡️" + " " + message
        } else if logDetails.message.contains("WebSocket Pong") {
            return "⬅️🏓" + " " + message
        } else {
            return message
        }
    }
}
```
add this as a new file to `Logger.Formatters` and use it in sample project

File: StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift (hunk 17 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift:282]: @VojtaStavik @b-onc 
I think this is the most important requested changes for this PR but it will required to rebuild some tests.
Should I do that? I think the current tests is more like component integration that cover all needed stuff.

File: Sources/Client/WebSocket/WebSocket.swift (hunk 18 above)
Comment (buh) [reply Sources/Client/WebSocket/WebSocket.swift:15]: oh no!

File: StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift (hunk 17 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift:282]: It's OK you need to rebuild the tests, because you moved the functionality out. The timing tests should be in the `PingController` tests, here should be just the tests for integration (as mentioned in my original comment).

Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift]: Fixed.

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController_Tests.swift [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        XCTAssertEqual(1, pingCount)
+        pingController.pongRecieved()
+        time.run(numberOfSeconds: WebSocketPingController.pongTimeoutTimeInterval)
+        XCTAssertEqual(0, reconnectCount)
+        
+        time.run(numberOfSeconds: WebSocketPingController.pingTimeInterval)
+        XCTAssertEqual(2, pingCount)
+        time.run(numberOfSeconds: WebSocketPingController.pongTimeoutTimeInterval)
+        XCTAssertEqual(1, reconnectCount)
+    }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController_Tests.swift:55]: Feel free to help the reader with some comments of what you're doing in the tests :)

File: StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+         timerQueue: DispatchQueue,
+         ping: @escaping () -> Void,
+         forceReconnect: @escaping () -> Void) {
+        self.timerType = timerType
+        self.timerQueue = timerQueue
+        self.ping = ping
+        self.forceReconnect = forceReconnect
+    }
+    
+    func connectionStateDidChange(_ connectionState: ConnectionState) {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketPingController.swift:45]: It'd be nice to add a comment that WSClient should call this when the connection state changes

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 15 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:92]: @b-onc , what's your opinion on this?

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 21]
```diff
@@ -94,6 +82,15 @@ class WebSocketClient {
     /// An object containing external dependencies of `WebSocketClient`
     private let environment: Environment
     
+    private(set) lazy var pingController: WebSocketPingController =
+        environment.createPingController(environment.timerType,
+                                         engineQueue,
+                                         { [weak self] in self?.engine.sendPing() },
+                                         { [weak self] in
+                                             self?.disconnect()
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:90]: I think we need to add a new disconnection reason - something like `noPongReceived`. This call uses the default value, which is `userInitiated`. That's actually all you had to do I think. The `reconnectIfNeeded` refactoring is redundant. The original implementation of `websocketDidDisconnect` would handle this correctly and start reconnecting automatically. It seems all we need to do here is actually just to call disconnect with the correct reason :magic:

File: StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift [hunk 22]
```diff
@@ -239,20 +241,42 @@ class WebSocketClient_Tests: StressTestCase {
...
+        XCTAssertEqual(engine.sendPing_calledCount, 1)
+        timeInterval += WebSocketPingController.pongTimeoutTimeInterval
+        time.run(numberOfSeconds: timeInterval)
+        XCTAssertEqual(engine.disconnect_calledCount, 1)
+        XCTAssertFalse(webSocketClient.connectionState.isConnected)
+        
+        AssertAsync {
+            Assert.willBeTrue(self.engine.connect_calledCount == 2)
+        }
     }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift:280]: None of this is needed, really :) You already test this functionality in the `WebSocketPingController_Tests`. You can remove these two tests and replace them with:
```
func test_webSocketPingController_ connectionStateDidChange_calledWhenConnectionChanges() {}
func test_webSocketPingController_ping_callsEngineWithPing()
func test_webSocketPingController_forceReconnect_disconnectsEngine()
```
All these tests should be very simple, synchronous, without any mention of virtual time. You just need to test is the integration between `WebSocketClinet` and `WebSocketPingController`, not the functionality of `WebSocketPingController`.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift [hunk 23]
```diff
@@ -537,3 +561,18 @@ class MockBackgroundTaskScheduler: BackgroundTaskScheduler {
...
+    override func connectionStateDidChange(_ connectionState: ConnectionState) {
+        connectionStateDidChangeCount += 1
+        super.connectionStateDidChange(connectionState)
+    }
+    
+    override func pongRecieved() {
+        pongRecievedCount += 1
+        super.pongRecieved()
+    }
+}
```
Comment (VojtaStavik) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient_Tests.swift:578]: Yes 👍

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 15 above)
Comment (b-onc) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:92]: Agreed, I think a delegate would be better (composition)

Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:92]: You want delegate because it’s unclear? A lot of code or what?
